lib/ABI/ApplyRegisterStateDeductions.cpp

```diff
@@ -0,0 +1,292 @@
+  auto RVAccessor = [](auto &FS) -> auto & { return FS.AsAnReturnValue; };
+
+  bool ArgRequired = false;
+  bool RVRequired = false;
+  for (auto [First, Second] : llvm::reverse(llvm::zip(GPAR, VAR))) {
+    LocalState FirstArgPair{ First };
+    LocalState FirstRVPair{ First };
+    if (auto Iterator = State.find(First); Iterator != State.end()) {
+      FirstArgPair.State = &Iterator->second.AsAnArgument;
+      FirstRVPair.State = &Iterator->second.AsAnReturnValue;
+    }
+
+    LocalState SecondArgPair{ Second };
+    LocalState SecondRVPair{ Second };
+    if (auto Iterator = State.find(Second); Iterator != State.end()) {
+      SecondArgPair.State = &Iterator->second.AsAnArgument;
+      SecondRVPair.State = &Iterator->second.AsAnReturnValue;
+    }
+
+    singlePositionBasedDeduction<CallingConvention,
+                                 EnsureABIConformance>(FirstArgPair,
+                                                       SecondArgPair,
+                                                       ArgRequired,
+                                                       State,
+                                                       ArgumentAccessor);
+
+    singlePositionBasedDeduction<CallingConvention,
+                                 EnsureABIConformance>(SecondArgPair,
+                                                       SecondRVPair,
+                                                       RVRequired,
+                                                       State,
+                                                       RVAccessor);
+  }
+
+  if constexpr (EnsureABIConformance == true) {
+    using CRegister = const model::Register::Values;
+    auto PossibleArguments = llvm::concat<CRegister>(GPAR, VAR);
+    auto PossibleRetValues = llvm::concat<CRegister>(GPRVR, VRVR);
+
+    // Mark all the non-conforming registers as `No`
+    for (auto &[Register, RegisterState] : State) {
+      auto &[AsAnArgument, AsAnReturnValue] = RegisterState;
+
+      if (model::RegisterState::isYesOrDead(AsAnArgument))
+        if (llvm::find(PossibleArguments, Register) == PossibleArguments.end())
+          AsAnArgument = model::RegisterState::No;
+
+      if (model::RegisterState::isYesOrDead(AsAnReturnValue))
+        if (llvm::find(PossibleRetValues, Register) == PossibleRetValues.end())
+          AsAnReturnValue = model::RegisterState::No;
+    }
+  }
+
+  return State;
+}
+
+template<typename CallingConvention, bool EnsureABIConformance = false>
+bool singleNonPositionBasedDeduction(model::RegisterState::Values &State,
+                                     bool IsRequired) {
+  if (model::RegisterState::isYesOrDead(State)) {
+    if (IsRequired) {
+      // Current register is already marked "Yes" or "Dead", do nothing.
+    } else {
+      // Current register is marked "Yes" or "Dead" but it shouldn't be.
+      if constexpr (EnsureABIConformance)
+        State = model::RegisterState::No;
```

Comment:
* Can we log when `EnsureABIConformance` does something that contradicts what was already there?
* This function would be nicer if it was to return a new value for `State`, rather than "tampering with" a reference.
* I'm thinking that, in `EnsureABIConformance`, we should drop all `Contradiction` and `Maybe`s. We should also change the default from `Maybe` to `No`. This change would require `RegisterStateMap` to embed a default value that we can change. But we can do this in the future I guess.

Reply:
1. Sure.
2. It works on a copy - the function higher down the line copies the state to be tampered with.
3. That actually sounds like a lovely idea.

Reply:
2. Yeah, I know, but having this function to be pure makes reading the caller side more straightforward, maybe. You could return a pair and use structured bindings?

Reply:
2. Actually, localizing the impact does sounds like a cool idea! Sorry I didn't think of that.

Reply:
Done, done and done (https://github.com/revng/revng/commit/f06daab3a2ce9a6ef1a4e357579c4cf7c16a57b8)